### ISSUE TITLE
option to only disable lint -> clippy::lint deprecation warnings

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#3159]
Some crates require to be compatible with stable.
They might have old-style clippy lint attributes.
Until the `tool_lints` features is available in stable, there will be a lot of 
````
warning: lint name `if_not_else` is deprecated and may not have an effect in the future. Also `cfg_attr(cargo-clippy)` won't be necessary anymore
  --> src/main.rs:16:52
   |
16 | #![cfg_attr(feature = "cargo-clippy", deny(clippy, if_not_else, enum_glob_use, wrong_pub_self_convention))]
   |                                                    ^^^^^^^^^^^ help: change it to: `clippy::if_not_else`

````
spam, so it would be cool to have a way to disable the `lint` -> `clippy::lint` transition warning explicitly until `tool_lints` hits stable, unless I am missing something and there already is a way to do that?
`renamed_and_removed_lints` implies that removed/unknown lints is still warned about which we still want warnings about in this case I think?

**Comment by flip1995** [comment]
Well I kind of waited for this issue to come up :smile: 

The only way is currently to `#[allow(renamed_and_removed_lints)]`. This is because this warning will only exist for a temporary time in the compiler (while the transition to `tool_lints` is made) and I didn't want to introduce a new lint in the compiler for that. 

I think `tool_lints` are nearly ready to be stabilized, the only thing that I don't currently like is the renamed/removed behavior of `tool_lints`. But I'll try to get this straight ASAP.

Until that point we can try to do two thinks
- create a new lint in the compiler explicitly for this warning (I'm not a fan of this)
- test in the compiler if the crate is compiled with stable toolchain and don't emit the warning in this case (not sure if this is possible, but I think it should be)

**Comment by Manishearth** [comment]
Why not continue to use cfg_attr with the new lint names for the time being? That way stuff still compiles.

**Comment by flip1995** [comment]
Oh yeah you're right that would be the best solution:
```rust
#![cfg_attr(feature="cargo-clippy", feature(tool_lints))]

#[cfg_attr(feature="cargo-clippy", warn(clippy::if_not_else))]
fn main() {
    if !true {
        println!("Hi, World!");
    } else {
        println!("Hello, World!");
    }
}
```

[Playground](https://play.rust-lang.org/?gist=c09e5045f97132e327da765985d3b78b&version=stable&mode=debug&edition=2015) compiles without errors or warnings and the Clippy warnings are emitted when using Clippy

**Comment by df5602** [comment]
When using the above trick, I start to get a warning in RLS in VSCode:

```[rustc] #![feature] may not be used on the beta release channel```

I can silence the RLS warning by changing the setting of the Rust extension to use RLS from stable instead of beta, but I'm not sure if the warning wouldn't just pop up again with the next stable...

I run clippy like that: ```cargo +nightly clippy```, so I tried changing the ```[cfg_attr(feature = "cargo-clippy", ...]``` to ```[cfg_attr(all(feature = "nightly", feature = "cargo-clippy"), ...]``` but that didn't work. (Or rather, it silenced the RLS warning, but the disabling of the clippy lint didn't work anymore..)

Any hints on how to solve this properly?

**Comment by mati865** [comment]
@df5602 your VS Code/Project is configured to use beta Rust.
That trick works for me in VS Code with nightly toolchain.

**Comment by df5602** [comment]
Can confirm. So that trick works in stable and nightly, but not beta?

(About the other thing with ```feature = "nightly"```: The blog where I copied this from defined this as a custom feature, but I used it assuming it was a built-in feature, similar to compilation target, etc.. So it's clear why that didn't work...)

**Comment by mati865** [comment]
@df5602 features are [considered unstable](https://doc.rust-lang.org/unstable-book/the-unstable-book.html) and can be used only with nightly toolchain.

**Comment by mati865** [comment]
@df5602 there is no `clippy-preview` for stable toolchain. Something that doesn't exist cannot produce error :wink:

**Comment by df5602** [comment]
Exactly ;-) I wasn't aware that RLS automatically used ```--feature="cargo-clippy"``` (on beta and nightly) behind the scenes. Now it all makes sense. Hopefully this is useful to know for others as well...

**Comment by mati865** [comment]
@flip1995 this will become more problematic because you cannot use features there.
I think deprecation notice should be disabled for beta branch once it's branched. Otherwise Clippy users with Rust 1.31 (when it comes out) will get get warning saying they should do something that is not possible on stable/beta.

**Comment by Arnavion** [comment]
Even for nightly users, it's weird that code that used to work suddenly gives warnings, *and* those warnings require enabling a whole new feature to silence.

**Comment by flip1995** [comment]
Yes I also think this is weird for nightly users... But I guess there is no good alternative. We can't just add a feature into stable and just hope that it works. And we _really_ want to move to `tool_lints`, because the whole `cfg_attr` thingies are pretty ugly IMO (and others), just to enable/disable a lint. So I guess a warning is the best thing to do.

---
@mati865  Will there be a `clippy-preview` on stable channel soon?

**Comment by phansch** [comment]
Going to close this as tool lints are on stable now :tada:
